File: ruby/iq14/stack.rb
#!/usr/bin/ruby
class Node
	attr_accessor :value, :bottom 
	def initialize(value)
		@value = value
		@bottom = nil
	end	
end	

class Stack
	def initialize(value)
		@top = Node.new(value)
	end	

	def push(value)
		temp = Node.new(value)
		temp.bottom = @top
		@top = temp
	end	

	def pop()
		if @top != nil
			value = @top.value
			@top = @top.bottom
			return value
		else
			puts "Stack is empty can't pop"	
		end	
	end	

	def peak()
		if @top != nil
			return @top.value
		else
		 	puts "Stack is empty"	
		end	
	end

	def printStack()
		cur = @top
		while(cur != nil)
			puts "------------"
			puts cur.value
			puts "------------"
			cur = cur.bottom
		end	
	end	
end	

Stack_one = Stack.new(1)
Stack_one.printStack()
Stack_one.push(2)
Stack_one.push(3)
Stack_one.push(4)
Stack_one.push(5)
Stack_one.push(6)
Stack_one.printStack()
Stack_one.pop()
Stack_one.printStack()

File: python/Stack_And_Queue_Questions/stack.py
#!/usr/bin/python

class Node(object):
  def __init__(self, value):
    self.value = value
    self.bottom = None

class Stack(object):
  def __init__(self):
    self.top = None
  def push(self, value):
    if self.top == None:
      self.top = Node(value)
    else:
      temp = Node(value)
      temp.bottom = self.top
      self.top = temp
  def pop(self):
    if self.top == None:
      return None
    else:
      temp = self.top.value
      self.top = self.top.bottom
      return temp

  def printStack(self):
    current = self.top
    while(current != None):
      print "| %s |" %current.value 
      print "------"
      current = current.bottom
    print " - - - - - - - - - - - - - "  
  
def main():
  testStack  = Stack()
  testStack.push(2)
  testStack.push(4)
  testStack.push(7)
  testStack.printStack()
  testStack.pop()
  testStack.printStack()
if __name__ == "__main__":
  main()     
